fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/util/pair/NonnullPair.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * NonnullPair.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.util.pair;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.annotation.SpotBugsSuppressWarnings;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A {@link Pair} implementation where both elements are not null. This utility allows
+ * the caller to omit null checks when retrieving elements from the pair.
+ *
+ * @param <L> the type of the left element
+ * @param <R> the type of the right element
+ */
+@API(API.Status.INTERNAL)
+public class NonnullPair<L, R> implements Pair<L, R> {
+    @Nonnull
+    private final Pair<L, R> pair;
```

Comment:
Can we omit this and just store left and right? We could avoid an extra allocation.

But If you intend to give the caller the ability to use an _existing_ `Pair` object as non-null, maybe you could make `NonnullPair` an interface, and provide a utility method in `Pair` that allows the user to treat _current_ `Pair` object as non-null:

```java
final var p = Pair.of(3, 4);
final var nnp = p.asNonnullable();
```

Or at least provide another constructor that accepts passing a `Pair` argument allowing it to strengthen its contract (making it non null).

Reply:
The main reason not to store `left` and `right` separately is that we'd need to basically duplicate the code in `ImmutablePair.java`. In theory, if we had templating, we could let the compiler do that, but not in vanilla Java. I'm not _too_ concerned with the extra allocation, because Hotspot may be able to do that inlining for us, and if it can't, I think this is a place where it would be better to see if there are performance problems first and then optimize.

The other reason concerns `MutablePair`s: there's not (currently) a `MutablePair` implementation, though we do use `MutablePair`s once in the codebase. (I may actually remove that usage in a future refactoring, in which case this becomes a bit moot.) But if you wanted a non-null AND mutable pair, then you could do this by wrapping a `MutablePair` in a `NonnullPair`. As you say, we'd also need an API for wrapping a `Pair`--and I think either of the ones you've suggeseted would do that.

I do think that those are API improvements we could consider. So far, in converting code, I haven't really needed them, but I can see their utility. My inclination is to add this as a follow up later if we need it.